resolve/hykkt/ruiz/RuizScalingKernelsCPU.cpp

```diff
@@ -0,0 +1,117 @@
+#include "RuizScalingKernelsCPU.hpp"
+
+namespace ReSolve
+{
+  using index_type = ReSolve::index_type;
+  using real_type  = ReSolve::real_type;
+
+  namespace hykkt
+  {
+    /**
+     * @brief CPU implementation of adaptRowMax. See RuizScalingKernelImpl.hpp.
+     *
+     * @param[in] n_hes - Number of rows in the Hessian matrix.
+     * @param[in] n_total - Total number of rows in the system.
+     * @param[in] hes_i - Row pointers for the Hessian matrix.
+     * @param[in] hes_j - Column indices for the Hessian matrix.
+     * @param[in] hes_v - Values for the Hessian matrix.
+     * @param[in] jac_i - Row pointers for the Jacobian matrix.
+     * @param[in] jac_j - Column indices for the Jacobian matrix.
+     * @param[in] jac_v - Values for the Jacobian matrix.
+     * @param[in] jac_tr_i - Row pointers for the transposed Jacobian matrix.
+     * @param[in] jac_tr_j - Column indices for the transposed Jacobian matrix.
+     * @param[in] jac_tr_v - Values for the transposed Jacobian matrix.
+     * @param[out] scaling_vector - Scaling vector to be updated.
+     */
+    void hykkt::RuizScalingKernelsCPU::adaptRowMax(index_type n_hes, index_type n_total, const index_type* hes_i, const index_type* hes_j, const real_type* hes_v, const index_type* jac_i, const index_type* jac_j, const real_type* jac_v, const index_type* jac_tr_i, const index_type* jac_tr_j, const real_type* jac_tr_v, real_type* scaling_vector)
+    {
+      for (index_type i = 0; i < n_hes; i++)
+      {
+        real_type max_l = 0;
+        real_type max_u = 0;
+        for (index_type j = hes_i[i]; j < hes_i[i + 1]; j++)
+        {
+          real_type entry = fabs(hes_v[j]);
+          if (entry > max_l)
+          {
+            max_l = entry;
+          }
+        }
+        for (index_type j = jac_tr_i[i]; j < jac_tr_i[i + 1]; j++)
+        {
+          real_type entry = fabs(jac_tr_v[j]);
+          if (entry > max_u)
+          {
+            max_u = entry;
+          }
+        }
+        if (max_l > max_u)
+        {
+          scaling_vector[i] = 1.0 / std::sqrt(max_l);
+        }
+        else
+        {
+          scaling_vector[i] = 1.0 / std::sqrt(max_u);
+        }
+      }
+      for (index_type i = n_hes; i < n_total; i++)
+      {
+        double max_l = 0;
+        for (index_type j = jac_i[i - n_hes]; j < jac_i[i - n_hes + 1]; j++)
+        {
+          real_type entry = fabs(jac_v[j]);
+          if (entry > max_l)
+          {
+            max_l = entry;
+          }
+        }
+        scaling_vector[i] = 1.0 / std::sqrt(max_l);
+      }
+    }
+
+    /**
+     * @brief CPU implementation of adaptDiagScale. See RuizScalingKernelImpl.hpp.
+     *
```

Comment:
If you want to make a doxygen reference, then please reference the class not the file.
```suggestion
     * @brief CPU implementation of adaptDiagScale. See RuizScalingKernelImpl.
     *
```